# 10x Experimental Scenarios
This document discusses proposed changes to the organization of sequencing run spreadsheets used by HIMC technicians running the 10x assay and the computational team performing pre-processing. This document heavily references the [10x documentation](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/using/mkfastq).

# Background
The relationship between **biological samples** (e.g. a cell suspension extracted from a single biological source, see [glossary](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/glossary)), **sequencing libraries** (e.g. the end product from a 10x chip lane), **BCL files** (e.g. generally multiplexed from several sequencing libraries), **FASTQs** (e.g. the product of de-multiplexing BCL files, sub-divided by lane and read), **Cell Ranger pre-processing runs** (e.g. can utilize several BCL files for a pooled run), and **Cell Ranger outputs** (e.g. Feature barcode matrix (FBM), TCR/BCR contigs) can be complicated. Two examples from the documentation are shown below.

### 2 Samples, 2 Seq-Libraries, 1 Seq-Run/Flowcell, 2 FBMs
![alt text](https://support.10xgenomics.com/img/mkfastq-1.png "")

"In this example, we have two 10x libraries (each processed through a separate Chromium chip channel) that are multiplexed on a single flowcell. Note that after running cellranger mkfastq, we run a separate instance of the pipeline on each library"

### 1 Sample, 1 Seq-Library, 2 Seq-Run/Flowcells, 1 FBM
![alt text](https://support.10xgenomics.com/img/mkfastq-2.png "")
"In this example, we have one 10x library sequenced on two flowcells. Note that after running cellranger mkfastq, we run a single instance of the pipeline on all the FASTQ files generated"

The common scenario of sequencing the same 'sequencing library' more than once (in the above example) is the reason why the 10x technicians are moving towards a FASTQ level organization (rather than a 'sequencing library' level).

# Proposed 10X Technician Spreadsheets: FASTQ, Sample, Library Features, Features
These are the proposed spreadsheets that will be used by the 10x techs (not all experiment-related columns are shown). Laura is working on making a new FASTQ-level spreadsheet - we are not considering the granular run-level and read-level FASTQs as separate FASTQs and in this document referring to this set as a "FASTQ" or "set of FASTQs". Below is the in-progress outline for this spreadsheet along with some example configurations:

### FASTQ-Level Spreadsheet
| FASTQs  | Loading Sample | Hashed Sample | 10x Lane ID | Library Type | BCL Run ID  | Processing Run  |   
|---|---|---|---|---|---|---|
| H1_GEX_BCL-1 | H1_GEX  | False  | 1  | GEX  | BCL-1 | CR-1 |

`FASTQs` is the name of the set of FASTQs generated by Cell Ranger Count - this is composed of the loading sample name appended with the `BCL Run ID`. This allows us to handle the common scenario where the same sequencing library (e.g. tube of liquid that can be aliquoted from) is sequenced more than once. `Loading Sample` is the name of the sample that is loaded into a 10x chip lane (e.g. can consist of several samples via hashing). `Hashed Sample` is a True/False value indicating whether hashing has been done (will be redundant with the `Loading Sample` naming convention). `10x Lane ID` will indicate the specific lane number a sample is loaded into - this is necessary for situations where the same sample is loaded into several lanes and in that case will become part of the FASTQs name. `Library Type` indicates what type of library is being prepared (e.g. GEX, ADT; however the chemistry and version may or may not be included ... GEX_5-prime in this column). `BCL Run ID` is the name of the BCL file the FASTQs will be put into. `Processing Run` is the name of the processing run that the data is being organized under - a processing run consists of the jobs necessary to convert BCL inputs into FBM and TCR/VDJ outputs). 

### Sample-Level Spreadsheet
| Sample name  | Loading Sample |  Sample-Meta-Data ... | HTO | Expected Cell Number | Library Features | 
|---|---|---|---|---|--|
| S1_GEX  | H1_GEX | ... | HTO-1  | 5000 | LF-1 |

A CITE-seq run sequencing-library will, in general, contain ADTs and HTOs. The list of features assocaited with this sequencing library are stored in a separate `Library Features Table` (see below and foreign keys under the `Library Features` column). For now, we are omitting the `Sample-Meta-Data` which will include Cell Ranger necessary information like organism, reference transcriptome etc. 

### Library Features Table
| Library Features| Index |
|---|---|
| HTO-1  | HTO-Index-1 |
| CD8_HIMC-1_Lot-1  | ADT-Index-1 |  

The Sample-Level Spreasheet above has links to Library Feature Tables - for convenience are stored as paired-columns rather than rows. The Library Features column will contain a list of feature names. The number of features varies from library to library so this spreadsheet is expected to have pairs of columns of varying lengths.

The `Library Features` feature names contain the following pieces of information in an underscore separated fashion: 1) official gene symbol of the measured protein (or the protein instead if desired - otherwise other names will be stored as aliaes), 2) the unique HIMC id (e.g. HIMC-1), and the lot number (e.g. lot-1). 

The `Index` column contains the oligo index IDs that can be looked up in a different `Features Table`.

# 10x Required CSVs

### Sample Sheet CSV
| Lane| Sample | Index |
|---|---|---|
| 1  | S1_GEX | SI-GA-A3 |

This table is in the format of the simple sample sheet for mkfastq. `Lane` refers to the 10x chip lane. `Sample` refers to our loading sample name (it will be appended with the BCL name automatically (I think) by mkfastq). `Index` refers to the oligos used to de-multiplex the BCL - in order to process custom oligos (e.g. ADT/HTO) we will pass the actual oligo index sequences in the `Index` column. 

### Libraries CSV
|  fastqs | sample  |  library_type |
|---|---|---|
| /path/to/fastqs/ | S1_GEX_BCL-1 | Gene Expression |
| /path/to/fastqs/ | S1_HTO_BCL-1 | Custom |

`fastqs` give a path to the demultiplexed FASTQ files (cannot have comma-delimiited paths) - more than one path requres an additional row. `sample` is the sampe name assigned in the mkfastq simple sample sheet. `library_type` is self explanatory.

### Feature Reference CSV

| id | name | read | pattern | sequence | feature_type | target_gene_id (optional) | target_gene_name (optional) |
|---|---|---|---|---|---|---|---|
|   |   |   |   |   |   |   |   |

This document is only necessary for feature barcode runs. `id` unique id for the feature (can't collide with gene name), `name` human readable feature name (e.g. gene name), `read` specifies which sequencing read contains the sequence (e.g. R2), `pattern` specifies how to extract seq from read, `feature_type` (e.g. custom), `target_gene_id` and `target_gene_name` are CRISPR specific and optional.

# Proposed Processing-Run Spreasheets
Here we are defining a `Processing-Run` to consist of a set of Cell Ranger jobs performed on one or more BCL files to obtain one or more FBMs and/or TCR/BCR sequence analysis. These are the spreadsheets that are planned to be used as inputs for a specific `Processing-Run` to the rebuilt pipeline. They will be similar to the input spreadsheets that are directly given as arguments to Cell Ranger Mkfastq and Count, but will contain the additional information required all arguments (e.g. expected cell count) as well as a layout of how to run the full `Processing-Run` set of jobs.

We need the following custom sample sheet in order to construct all the arguments for Cell Ranger Count:

### Sample Sheet CSV
| Lane| Sample | Index | Library Type | Reference Transcriptome | Number of Cells |
|---|---|---|---|---|---|
| 1  | S1_GEX | SI-GA-A3 | Gene Expression | GRCh38 | 3000 |

# Enumeration of Scenarios

### scRNA-seq: 3 Samples, 3 10x Lanes, 3 Seq-Libraries, 1 Flowcell/BCL
These three rows represents an experiment (e.g. Cell Ranger Run `CR-1`) that has three samples (`S1`, `S2`, `S3`) run in separate 10x chip lanes. The three libraries generated from the three lanes are multiplexed and run in a single flowcell, which generates a single BCL file (`BCL-1`). This single BCL file will need to be de-multiplexed, producing three sets of FASTQs that will produce three 
feature-barcode-matrices (FBMs). 

| Library  | Sample  | BCL  | Cell Ranger Run  |   
|---|---|---|---|
| S1_GEX  | S1  | BCL-1  | CR-1  |   
| S2_GEX  | S2  | BCL-1  | CR-1  |   
| S3_GEX  | S3  | BCL-1  | CR-1  |  

```
BCL-1 -> FASTQs-1 -> FBM-1
      -> FASTQs-2 -> FBM-2
      -> FASTQs-3 -> FBM-3
```

### CITE-seq and Hashtagging: 2 Samples, 1 10x Lane, 3 Seq-Libraries (GEX, ADT, HTO), 1 Flowcell/BCL
The second set of three rows represents a CITE-seq and hashtag experiment (`CR-2`) that has two samples (`S4`, `S5`) run in one 10x chip lane, which produces three libraries (`S4_GEX`, `S4_ADT`, `S4_HTO`)

| Library  | Sample  | BCL  | Processing Run  |  Hashed |
|---|---|---|---| -- |
| S_GEX  | S4_S5  | BCL-2  | CR-2  |   |
| S_ADT  | S4_S5  | BCL-2  | CR-2  |   |
| S_HTO  | S4_S5  | BCL-2  | CR-2  |    |

```   
mkfastq              count     de-hashtag
BCL-2 -> FASTQs_GEX -> FBM -> FBM_S4
      -> FASTQs_ADT           FBM_S5
      -> FASTQs_HTO    
```

### scRNA-seq (pooled run): One Sample, Three Seq-Libraries, Three Flowcell/BCLs

| Library  | Sample  | BCL  | Cell Ranger Run  |   
|---|---|---|---| 
| S6_GEX_A  | S6  | BCL-3  | CR-3  |
| S6_GEX_B  | S6  | BCL-4  | CR-3  |   
| S6_GEX_C  | S6  | BCL-5  | CR-3  |  




